cellarium/ml/utilities/data.py

```diff
@@ -86,26 +95,28 @@ def collate_fn(batch: list[dict[str, np.ndarray]]) -> dict[str, np.ndarray | tor
         the batch dimension.
     """
     keys = batch[0].keys()
-    collated_batch: dict[str, np.ndarray | torch.Tensor] = {}
-    if len(batch) > 1:
-        if not all(keys == data.keys() for data in batch[1:]):
-            raise ValueError("All dictionaries in the batch must have the same keys.")
+    collated_batch: dict[str, dict[str, np.ndarray] | np.ndarray] = {}
+    if len(batch) > 1 and not all(keys == data.keys() for data in batch[1:]):
+        raise ValueError("All dictionaries in the batch must have the same keys.")
     for key in keys:
         if key.endswith("_g") or key.endswith("_categories"):
             # Check that all values are the same
             if len(batch) > 1:
-                if not all(np.array_equal(batch[0][key], data[key]) for data in batch[1:]):
+                if not all(np.array_equal(batch[0][key], data[key]) for data in batch[1:]):  # type: ignore[arg-type]
                     raise ValueError(f"All dictionaries in the batch must have the same {key}.")
             # If so, just take the first one
             value = batch[0][key]
+        elif isinstance(batch[0][key], dict):
+            subkeys = batch[0][key].keys()  # type: ignore[union-attr]
+            if len(batch) > 1 and not all(subkeys == data[key].keys() for data in batch[1:]):  # type: ignore[union-attr]
+                raise ValueError(f"All '{key}' sub-dictionaries in the batch must have the same subkeys.")
+            value = {subkey: np.concatenate([data[key][subkey] for data in batch], axis=0) for subkey in subkeys}
```

Comment:
Check that subkeys have the self batch dimension and otherwise, emit an informative error message.